### PR TITLE
Part 2 robot simulator

### PR DESCRIPTION
**Pull request by JoelChanZhiYang**
# Description
The EV3 robot simulator implementation is in PR. This looks like a big PR but about half of it is tests. 

image

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?
There is 119 tests to test the main aspects of the robot simulation. 

image

To try out the robot simulation. You can try out this snippet. 

In this snippet, you will observe

- Wheels
- Motors
- Paper and url loading
- Ultrasonic sensor
- ev3_pause
- Floor and Wall

```
import {
  ev3_motorA,
  ev3_motorB,
  ev3_motorC,
  ev3_motorD,
  ev3_runToRelativePosition,
  ev3_colorSensorRed,
  ev3_colorSensorGreen,
  ev3_pause,
  ev3_colorSensorBlue,
  ev3_ultrasonicSensor,
  ev3_ultrasonicSensorDistance,
  init_simulation,
  createRenderer,
  createPhysics,
  createTimer,
  createWorld,
  createPaper,
  createWall,
  createEv3,
  createFloor,
  createCSE,
  createRobotConsole,
  addControllerToWorld,
  saveToContext
} from "robot_simulation";


init_simulation(() => {
    const physics = createPhysics();
    const renderer = createRenderer();
    const timer = createTimer();
    const robot_console = createRobotConsole();
    
    const world = createWorld(physics, renderer, timer, robot_console);
    
    const program = createCSE();
    addControllerToWorld(program, world);
    
    const paper = createPaper(
        renderer, 
        "https://www.shutterstock.com/image-vector/black-white-stripes-260nw-785326606.jpg", 
        1, 
        1);
        
    addControllerToWorld(paper, world);
    
    const floor = createFloor(physics, renderer);
    addControllerToWorld(floor, world);
    
    const wall = createWall(physics, renderer);
    addControllerToWorld(wall, world);
    
    const ev3 = createEv3(physics, renderer);
    addControllerToWorld(ev3, world);
    
    saveToContext("world", world);
    saveToContext("ev3", ev3);
    
    return world;
});

// Student code below

let motorA = ev3_motorA();
let motorB = ev3_motorB();
let ultrasonicSensor = ev3_ultrasonicSensor();

ev3_pause(2000);

let loop = true;

while (loop) {
    let distance = ev3_ultrasonicSensorDistance(ultrasonicSensor);
    
    if (distance < 50) {
        loop = false;
        continue;
    }
    
    ev3_runToRelativePosition(motorA, 360, 200);
    ev3_runToRelativePosition(motorB, 360, 200);
    ev3_pause(3000);
}
```

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/jest.config.js [hunk 1]
```diff
@@ -20,7 +20,8 @@ export default {
   moduleNameMapper: {
     // Module Name settings required to make chalk work with jest
     '#(.*)': '<rootDir>/node_modules/$1',
-    '^js-slang/context': '<rootDir>/__mocks__/context.ts'
+    '^js-slang/context': '<rootDir>/__mocks__/context.ts',
+    '^three.+.js': '<rootDir>/__mocks__/emptyModule.js',
```
Comment (JoelChanZhiYang) [inline src/jest.config.js:24]: This is a fix to allow jest to understand the untranspiled Three examples like OrbitalControl and GLTFLoader

https://stackoverflow.com/questions/58363166/errors-cannot-use-import-statement-outside-a-module-three-orbitcontrol

File: src/tabs/RobotSimulation/components/TabPanels/ConsolePanel.tsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    );
+  }
+
+  return (
+    <TabWrapper>
+      <LastUpdated time={timing} />
+      <ul>
+        {logs.map((log, i) => (
+          <li key={i}>{getLogString(log)}</li>
+        ))}
```
Comment (RichDom2185) [inline src/tabs/RobotSimulation/components/TabPanels/ConsolePanel.tsx:56]: Can individual logs be cleared? E.g. is the below scenario possible?

Initial: log entries with timestamps 1-5

then delete log entry 4

Next state: `logs` only contain entries 1, 2, 3, 5.

---

If yes, using the index as key will cause rendering bugs as the some entries may not update.

File: src/bundles/robot_simulation/controllers/program/evaluate.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
+
```
Comment (RichDom2185) [inline src/bundles/robot_simulation/controllers/program/evaluate.ts:1]: There seems to be an extra blank line here.

File: src/bundles/robot_simulation/controllers/ev3/ev3/default/types.ts
Comment (RichDom2185) [inline src/bundles/robot_simulation/controllers/ev3/ev3/default/types.ts:1]: Can we fix the indent warnings in this file?

**Review by RichDom2185 — COMMENTED** [review]
Hi, thanks for working on this! Very excited to try it out; have not looked through all the files but in the meantime, here are some preliminary comments:

File: yarn.lock [hunk 4]
```diff
@@ -385,6 +385,11 @@
   resolved "https://registry.yarnpkg.com/@dimforge/rapier3d-compat/-/rapier3d-compat-0.11.2.tgz#ae2b335f545decf1e82ff45bb10368e143de0fcb"
   integrity sha512-vdWmlkpS3G8nGAzLuK7GYTpNdrkn/0NKCe0l1Jqxc7ZZOB3N0q9uG/Ap9l9bothWuAvxscIt0U97GVLr0lXWLg==
 
+"@dimforge/rapier3d-compat@^0.11.2":
+  version "0.11.2"
+  resolved "https://registry.yarnpkg.com/@dimforge/rapier3d-compat/-/rapier3d-compat-0.11.2.tgz#ae2b335f545decf1e82ff45bb10368e143de0fcb"
+  integrity sha512-vdWmlkpS3G8nGAzLuK7GYTpNdrkn/0NKCe0l1Jqxc7ZZOB3N0q9uG/Ap9l9bothWuAvxscIt0U97GVLr0lXWLg==
+
```
[the remark this answers is not in the document]
Comment (JoelChanZhiYang) [reply yarn.lock:392]: You are right. Fixed!

File: src/bundles/robot_simulation/controllers/program/evaluate.ts (hunk 3 above)
Comment (JoelChanZhiYang) [reply src/bundles/robot_simulation/controllers/program/evaluate.ts:1]: Fixed

File: src/tabs/RobotSimulation/components/TabPanels/UltrasonicSensorPanel.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    return ultrasonicSensor.sense();
+  }, 1000);
+
+  if (timing === null) {
+    return <TabWrapper>Loading color sensor</TabWrapper>;
+  }
+
+  if (distanceSensed === null) {
+    return <TabWrapper>Color sensor not found</TabWrapper>;
+  }
```
[the remark this answers is not in the document]
Comment (JoelChanZhiYang) [reply src/tabs/RobotSimulation/components/TabPanels/UltrasonicSensorPanel.tsx:25]: Fixed

File: src/bundles/robot_simulation/controllers/ev3/ev3/default/config.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    width: 0.145,
+    length: 0.18,
+  },
+  mass: 0.6,
+  type: 'dynamic',
+  debug: true,
+};
+
+export const meshConfig: Ev3MeshConfig = {
+  url: 'https://keen-longma-3c1be1.netlify.app/6_remove_wheels.gltf',
```
Comment (RichDom2185) [inline src/bundles/robot_simulation/controllers/ev3/ev3/default/config.ts:40]: Are we going to eventually move this into a/the repo?

File: src/bundles/robot_simulation/controllers/ev3/ev3/default/config.ts [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+      derivativeGain: 0,
+      integralGain: 0,
+    },
+    mesh: {
+      dimension: {
+        width: 0.028,
+        height: 0.0575,
+        length: 0.0575,
+      },
+      url: 'https://keen-longma-3c1be1.netlify.app/6_wheel.gltf',
```
Comment (RichDom2185) [inline src/bundles/robot_simulation/controllers/ev3/ev3/default/config.ts:96]: Ditto.

File: src/bundles/robot_simulation/controllers/program/Program.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        throw Error('Program not started');
+      }
+
+      if (this.isPaused) {
+        return;
+      }
+
+      // steps per tick
+      for (let i = 0; i < this.config.stepsPerTick; i++) {
+        const result = this.iterator.next();
```
Comment (RichDom2185) [inline src/bundles/robot_simulation/controllers/program/Program.ts:70]: Is this unused variable intentional?

**Review by RichDom2185 — APPROVED** [review]
Hi, sorry for the delayed review, thanks a lot for working on this! LGTM, though I just have a few minor comments; would kindly ask if you could reply to them.

Thanks and it looks really good too! 🤩

File: src/bundles/robot_simulation/controllers/ev3/ev3/default/config.ts (hunk 6 above)
Comment (JoelChanZhiYang) [reply src/bundles/robot_simulation/controllers/ev3/ev3/default/config.ts:40]: Hello! Sorry for the slow reply! 

How do other modules host their static files? Is it inside the modules repo or in a separate repo?

Comment (RichDom2185) [reply src/bundles/robot_simulation/controllers/ev3/ev3/default/config.ts:40]: So far I think the only other module with static assets is unity academy, and we host the files in an S3 bucket on AWS

Comment (JoelChanZhiYang) [reply src/bundles/robot_simulation/controllers/ev3/ev3/default/config.ts:40]: Ah alright. How do I go about to do that?

Comment (RichDom2185) [reply src/bundles/robot_simulation/controllers/ev3/ev3/default/config.ts:40]: Let's take this over to Telegram.

**Pull request merged**